README.md
#RFID-Chafon

I recently bought a usb rfid reader. The "driver" and the "app" were propietaries solutions for windows xp and not usable by mac or linux. So I decided to reverse engineer the protocol and implement this small node js lib.

## Installation

```bash
$ npm install debug
```

## Usage


Example _readtag.js_:

```js
var RFID = require('./index');

let rfid = new RFID('/dev/tty.usbserial');

rfid.open()
	.then(() => {
		return rfid.read().then(result => {
			console.log(result);
			rfid.close();
		});
	}).catch(e => {
		console.error(e);
		rfid.close();
	});

```

Example _writetag.js_:

```js
var RFID = require('./index');

let rfid = new RFID('/dev/tty.usbserial');

rfid.open()
	.then(() => {
		return rfid.write().then(result => {
			console.log(result);
			rfid.close();
		});
	}).catch(e => {
		console.error(e);
		rfid.close();
	});

```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.1.4](https://github.com/ulfalfa/rfid-chafon/compare/v2.1.3...v2.1.4) (2019-05-25)


### Bug Fixes

* **chafon:** return empty reads with undefined and throw crc error ([ec51528](https://github.com/ulfalfa/rfid-chafon/commit/ec51528))



### [2.1.3](https://github.com/ulfalfa/rfid-chafon/compare/v2.1.2...v2.1.3) (2019-05-24)


### Bug Fixes

* **d302:** removing all listeners in case of timeout ([55a2a57](https://github.com/ulfalfa/rfid-chafon/commit/55a2a57))



### [2.1.2](https://github.com/ulfalfa/rfid-chafon/compare/v2.1.1...v2.1.2) (2019-05-24)


### Bug Fixes

* **d302:** replaced listening with ```on``` with ```once````to prevent mem leak ([597538d](https://github.com/ulfalfa/rfid-chafon/commit/597538d))



### [2.1.1](https://github.com/ulfalfa/rfid-chafon/compare/v2.1.0...v2.1.1) (2019-05-24)



## [2.1.0](https://github.com/ulfalfa/rfid-chafon/compare/v2.0.1...v2.1.0) (2019-05-24)


### Build System

* added release command ([1dc167d](https://github.com/ulfalfa/rfid-chafon/commit/1dc167d))


### Features

* **rfid:** interval observe (periodical readings) ([3b24205](https://github.com/ulfalfa/rfid-chafon/commit/3b24205))



### [2.0.1](https://github.com/ulfalfa/rfid-chafon/compare/v2.0.0...v2.0.1) (2019-05-22)



## [2.0.0](https://github.com/ulfalfa/rfid-chafon/compare/v1.1.1...v2.0.0) (2019-05-22)


### Features

* rewritten with typescript and added second reader type ([bf81cd1](https://github.com/ulfalfa/rfid-chafon/commit/bf81cd1))


### BREAKING CHANGES

* complete rewrite


test.js
var RFID = require('./dist/index').RFID;

let rfid = new RFID('/dev/tty.usbserial');

/*rfid.open()
	.then((result) => {

		return rfid.write('AA345678AA').then(result => {
			console.log(result);
			rfid.close();
		});

	}).catch(e => {
		console.error(e);
		rfid.close();
	});
	*/

rfid.on('data', (...data) => {
	console.log('Tag read', data);
});
rfid.on('error', (data) => {
	console.log('Error', data);
});
rfid.startReading(1);


setTimeout(() => {
	rfid.close();

}, 10000);

src/d302.spec.ts
import { serial, TestInterface } from 'ava'

import { D302Reader as RFID } from './d302'
import { tap, map, take } from 'rxjs/operators'

interface Context {
  rfid: RFID
}

const test = serial as TestInterface<Context>

const PORT = '/dev/tty.wchusbserial14320'

test.beforeEach(async t => {
  const rfid = new RFID(PORT)
  await rfid.open()

  t.context = {
    rfid,
  }
})

test.afterEach.always(async t => {
  await t.context.rfid.close()
})

test('can get Info', async t => {
  const rfid = t.context.rfid

  const result = await rfid.getInfo()
  t.is(result, 'OK')
})

test('observe', t => {
  t.plan(10)

  const rfid = t.context.rfid

  return rfid.observe().pipe(
    take(10),
    tap(data => {
      t.log(data)
      t.pass()
    })
  )
})

test('can read tag', async t => {
  const rfid = t.context.rfid

  const result = await rfid.read()
  t.log(result)
  t.pass()
})

test('write and read', async t => {
  const rfid = t.context.rfid

  await rfid.write('012345679A', { safe: false })

  const result = await rfid.read()
  t.is(result, '012345679A')
})
test('safe write with default', async t => {
  const rfid = t.context.rfid

  const result = await rfid.write('012345679B')
  t.true(typeof result === 'undefined')
})

test('safe write', async t => {
  const rfid = t.context.rfid

  const result = await rfid.write('012345679B', { safe: true })
  t.true(typeof result === 'undefined')
})


src/rfid.spec.ts
import anyTest, { serial, Macro, TestInterface } from 'ava'

import { RFID } from './rfid'
import { tap, map, take } from 'rxjs/operators'
const test = anyTest

test.only('Get Info', async t => {
  const rfid = new RFID('/dev/tty.usbserial')

  await rfid.open()
  const result = await rfid.getInfo()
  t.log(Buffer.from(result, 'utf8'))
  t.log(Buffer.from('ID card reader & writer6', 'ascii'))
  t.log(Buffer.from('ID card reader & writer6', 'ucs2'))
  t.log(Buffer.from('ID card reader & writer6', 'binary'))
  t.is(result, 'ID card reader & writer6')
  // t.is(result, 'ID card reader & writer6')

  // ulf 010C93215B
  // jutta 010C932300
})

test('Connect', async t => {
  const rfid = new RFID('/dev/tty.usbserial')

  await rfid.open()
  let result
  // await rfid.beep()
  result = await rfid.atomicWrite('010C93215B')
  // t.log(result)
  result = await rfid.read()
  t.log(result)
  t.is(result, '010C93215B')
  // t.is(result, 'ID card reader & writer6')

  // ulf 010C93215B
  // jutta 010C932300
})

test('Observe', t => {
  const rfid = new RFID('/dev/tty.usbserial')
  t.plan(10)
  return rfid.observe().pipe(
    take(10),
    map(data => {
      console.log(data)
      return t.pass()
    })
  )
})


index.js
/*jshint node: true*/
'use strict';


const SerialPort = require('serialport'),
	debug = require('debug')('rfid-chafon'),
	EventEmitter = require('events');

const rfidReaderParser = function () {

	var tmpBuf = new Buffer(50);
	var byteCount = 0;
	var pointer = 0;
	var csum = 0;
	var data = null;
	var skipNext = false;
	return function (emitter, inputBuf) {
		var pos = 0;
		var buffer;
		//we need at least 8 byte, before we start the real parsing (sometimes only 6 bytes arriving)
		if ((inputBuf.length + byteCount + pointer) < 4) {
			inputBuf.copy(tmpBuf, byteCount);
			byteCount = byteCount + inputBuf.length;

		} else {
			if (byteCount > 0) {
				buffer = Buffer.concat([tmpBuf.slice(0, byteCount), inputBuf]);
				byteCount = 0;

			} else {
				buffer = inputBuf;
			}

			if (!data) {
				data = {};
				data.id = buffer.readUInt16BE(pos).toString(16);
				pos = pos + 2;
				data.length = buffer.readUInt16BE(pos);
				pos = pos + 2;
				data.data = new Buffer(data.length);
				csum = 0;
			}
			var skip = 0;
			for (var i = pos; i < buffer.length; i++) {

				csum = csum ^ buffer[i];
				if (skipNext && buffer[i].toString(16) === '0') {
					skipNext = false;
				} else {
					data.data[pointer] = buffer[i];
					pointer++;
				}

				if (buffer[i].toString(16) === 'aa') {
					skipNext = true;
				}
			}

			if (pointer >= data.length) {
				pointer = 0;
				data.crc = (csum === 0);

				debug('Data received', data);
				emitter.emit('data', data);
				data = null;

			}
		}
	};
};



class RFID extends EventEmitter {

	constructor(port = '/dev/ttyUSB0') {

		super();
		this._readInterval = null;

		this.serialPort = new SerialPort(port, {
			baudrate: 38400,
			dataBits: 8,
			stopBits: 1,
			parity: 'none',
			flowControl: false,
			xon: false,
			xoff: false,
			rtscts: false,
			bufferSize: 30,
			parser: rfidReaderParser(),
			autoOpen: false

		}); // this is the openImmediately flag [default is true]

	}

	beep() {
		return new Promise((resolve, reject) => {

			if (!this.serialPort.isOpen()) {
				reject('not ready');
				return;
			}

			var cmd = new Buffer('aadd000401030A08', 'hex');
			this.serialPort.once('data', resolve);

			this.serialPort.write(cmd, function (err, result) {
				if (err) {
					reject(err);
				}

			});
		});
	}

	read() {

		return new Promise((resolve, reject) => {
			if (!this.serialPort.isOpen()) {
				reject('not ready');
				return;
			}
			var cmd = new Buffer('AADD0003010C0D', 'hex');
			this.serialPort.once('data', function (data) {
				debug('data read', data);
				if (!data.crc) {
					reject('CRCERROR');

				} else if (data.data[2] === 1) {
					resolve(undefined);

				} else {

					resolve(data.data.slice(3, 8).toString('hex').toUpperCase());


				}

			});
			this.serialPort.write(cmd, (err, result) => {
				if (err) {
					reject(err);
				}

			});
		});


	}

	atomicWrite(id) {


		return new Promise((resolve, reject) => {
			if (!this.serialPort.isOpen()) {
				reject('not ready');
				return;
			}
			if (typeof (id) !== 'string' || id.length !== 10) {
				reject('Wrong ID ' + id + ' length:' + id.length);
			} else {
				var cmdBuffer = new Buffer(50);
				cmdBuffer.write('AADD0009', 'hex');
				var pos = 4;
				//now write the compose the id for checksum
				var idBuffer = new Buffer(8);
				idBuffer.write('030C00', 'hex');
				idBuffer.write(id, 3, 'hex');

				debug('idBuffer', idBuffer);

				//now calculate checksum and fill cmdBuffer and doubling the 0xAA
				var csum = 0;
				for (var i = 0; i < idBuffer.length; i++) {
					//jshint bitwise:false
					csum = csum ^ idBuffer[i];
					cmdBuffer[pos] = idBuffer[i];
					pos++;
					//double the AAs
					if (idBuffer[i].toString(16) === 'aa') {
						cmdBuffer[pos] = idBuffer[i];
						pos++;
					}

				}

				cmdBuffer[pos] = csum;
				pos++;
				this.serialPort.once('data', resolve);
				debug('writing', cmdBuffer.slice(0, pos));
				this.serialPort.write(cmdBuffer.slice(0, pos), function (err, result) {
					if (err) {
						reject(err);
					}

				});

			}
		});

	}

	write(id) {
		return this.atomicWrite(id)
			.then(() => {
				return this.read();
			})
			.then((data) => {
				debug('Compare data', id, data);
				if (data && data === id.toUpperCase()) {
					return (data);

				} else {
					if (!data) {
						throw (new Error('NOTAG'));
					} else {
						throw (new Error('Unknown Error'));

					}
				}
			});
	}

	getInfo() {
		return new Promise((resolve, reject) => {
			if (!this.serialPort.isOpen()) {
				reject('not ready');
				return;
			}
			var cmd = new Buffer('AADD0003010203', 'hex');
			this.serialPort.once('data', function (data) {
				resolve(data.data.slice(3).toString());
			});
			this.serialPort.write(cmd, function (err, result) {
				if (err) {
					reject(err);
				}

			});
		});

	}


	open() {
		return new Promise((resolve, reject) => {
			this.serialPort.open((err) => {
				if (err) {
					reject(err);
				} else {
					resolve();
				}
			});
		});
	}

	close() {
		this.serialPort.close();

	}

	startReading(interval = 5) {
		return this.open()
			.then(() => {

				this._readInterval = setInterval(() => {
					this.read()
						.then(data => {
							debug('data', data);
							this.emit('data', data);
						}).catch(e => {
							this.emit('error', e);
						});

				}, interval * 1000);
			});

	}

	stopReading() {
		if (this._readInterval) {
			clearInterval(this._readInterval);
		}

	}

}

module.exports = exports = RFID;


src/rfid.ts
import * as SerialPort from 'serialport'

import { Observable, interval } from 'rxjs'
import { switchMap, filter, distinctUntilChanged } from 'rxjs/operators'

export abstract class RFIDReader {
  protected serialPort: SerialPort

  get isOpen(): boolean {
    return this.serialPort.isOpen
  }

  constructor(port = '/dev/ttyUSB0', baudRate: number) {
    this.serialPort = new SerialPort(port, {
      baudRate,
      dataBits: 8,
      stopBits: 1,
      parity: 'none',
      // flowControl: false,
      xon: false,
      xoff: false,
      rtscts: false,
      // bufferSize: 30,
      // parser: rfidReaderParser(),
      autoOpen: false,
    }) // this is the openImmediately flag [default is true]
  }

  abstract getInfo(): Promise<string>

  protected abstract atomicWrite(code: string): Promise<void>
  abstract read(): Promise<string>

  async write(
    code: string,
    options: { safe: boolean } = { safe: true }
  ): Promise<void> {
    if (code.length !== 10) {
      throw new Error(`Length code ${code} expected 10 but is ${code.length}`)
    }

    return this.atomicWrite(code).then(() => {
      if (options.safe) {
        return this.read().then(readCode => {
          if (readCode !== code) {
            throw new Error(
              `Code ${code} was not written (current ${readCode})`
            )
          }
        })
      }
    })
  }

  observe(intervalSec: number = 1): Observable<string> {
    return interval(intervalSec * 1000).pipe(
      switchMap(() => {
        return this.read().catch(e => {
          // do nothing
          return undefined
        })
      }),
      filter(data => !!data)
    )
  }

  open(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.serialPort.open(err => {
        if (err) {
          reject(err)
        } else {
          resolve()
        }
      })
    })
  }

  close(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.serialPort.close(err => {
        if (err) {
          reject(err)
        } else {
          resolve()
        }
      })
    })
  }
}


src/chafon.ts
import { parsers } from 'serialport'
import * as Debug from 'debug'
import { EventEmitter } from 'events'

import { RFIDReader } from './rfid'

import { Transform, TransformCallback } from 'stream'
import { timingSafeEqual } from 'crypto'

const CMD_INFO = 'AADD0003010203'

const CMD_BEEP = 'AADD000401030A08'

const CMD_READ = 'AADD0003010C0D'

const debug = Debug('ulfalfa:rfid:chafon')
const debug2 = Debug('ulfalfa:rfid:chafon:parser')

interface ChafonResult {
  data: Buffer
  id: string
  length: number
  crc: boolean
}

class ChafonParser extends Transform {
  buffer = Buffer.alloc(50)
  byteCount = 0
  constructor() {
    super({ objectMode: true })
  }

  parse() {
    const id = this.buffer.readUInt16BE(0).toString(16)
    const length = this.buffer.readUInt16BE(2)

    let csum
    let skipNext = false
    let pointer = 0
    const data = Buffer.alloc(50)
    for (let i = 4; i < this.buffer.length; i++) {
      // tslint:disable-next-line
      csum = csum ^ this.buffer[i]
      if (skipNext && this.buffer[i].toString(16) === '0') {
        skipNext = false
      } else {
        data[pointer] = this.buffer[i]
        pointer++
      }

      if (this.buffer[i].toString(16) === 'aa') {
        skipNext = true
      }
    }

    debug2(
      `Message ${id} with length ${length} arrived (CSUM:${csum === 0}) = `,
      data
    )
    this.push({ data: data.slice(0, length), id, length, crc: csum === 0 })
  }

  _transform(
    chunk: Buffer,
    encoding: string,
    callback: TransformCallback
  ): void {
    debug2(
      `Chunk arrived with ${chunk.length} byte, we alreay have ${
        this.byteCount
      }`,
      chunk
    )

    chunk.copy(this.buffer, this.byteCount)
    this.byteCount += chunk.length

    if (
      this.byteCount > 4 &&
      this.buffer.readUInt16BE(2) === this.byteCount - 4
    ) {
      debug2('Message complete')
      this.parse()

      this.byteCount = 0
      this.buffer = Buffer.alloc(50)
    }

    callback()
  }
}

export class ChafonReader extends RFIDReader {
  parser: ChafonParser

  constructor(port = '/dev/ttyUSB0') {
    super(port, 38400)

    this.parser = new ChafonParser()
    this.serialPort.pipe(this.parser)
  }

  public doCommand(cmdString: string | Buffer): Promise<ChafonResult> {
    const cmd =
      typeof cmdString === 'string' ? Buffer.from(cmdString, 'hex') : cmdString
    return new Promise((resolve, reject) => {
      this.parser.once('data', resolve)

      debug('>Send Command', cmd)
      this.serialPort.write(cmd, (err, result) => {
        if (err) {
          reject(err)
        }
      })
    }).then(data => {
      debug('< returned', data)
      return data as ChafonResult
    })
  }

  getInfo(): Promise<any> {
    return this.doCommand(CMD_INFO).then(data => {
      return data.data.slice(3).toString()
    })
  }

  beep() {
    return this.doCommand(CMD_BEEP)
  }

  read(): Promise<string> {
    debug('Entered Read')

    return this.doCommand(CMD_READ).then(data => {
      if (!data.crc) {
        throw new Error('CRCERROR')
      } else if (data.data[2] === 1) {
        return undefined
      } else {
        return data.data
          .slice(3, 8)
          .toString('hex')
          .toUpperCase()
      }
    })
  }

  protected atomicWrite(id: string): Promise<void> {
    debug('Entered AtomicWrite')

    const cmdBuffer = Buffer.alloc(50)
    cmdBuffer.write('AADD0009', 'hex')
    let pos = 4
    // now write the compose the id for checksum
    const idBuffer = Buffer.alloc(8)
    idBuffer.write('030C00', 'hex')
    idBuffer.write(id, 3, 'hex')

    // now calculate checksum and fill cmdBuffer and doubling the 0xAA
    let csum = 0
    for (const part of idBuffer) {
      // tslint:disable-next-line
      csum = csum ^ part
      cmdBuffer[pos] = part
      pos++
      // double the AAs
      if (part.toString(16) === 'aa') {
        cmdBuffer[pos] = part
        pos++
      }
    }

    cmdBuffer[pos] = csum
    pos++

    return this.doCommand(cmdBuffer.slice(0, pos)).then(data => {
      debug('Write Result', data)
      return
    })
  }
}


src/rfid-transform.ts
import { Transform, TransformCallback } from 'stream'
import * as Debug from 'debug'

const debug = Debug('rfid-chafon:transform')

export class ChafonParser extends Transform {
  buffer = Buffer.alloc(50)
  byteCount = 0
  constructor() {
    super({ objectMode: true })
  }

  parse() {
    const id = this.buffer.readUInt16BE(0).toString(16)
    const length = this.buffer.readUInt16BE(2)

    let csum
    let skipNext = false
    let pointer = 0
    const data = Buffer.alloc(50)
    for (let i = 4; i < this.buffer.length; i++) {
      // tslint:disable-next-line
      csum = csum ^ this.buffer[i]
      if (skipNext && this.buffer[i].toString(16) === '0') {
        skipNext = false
      } else {
        data[pointer] = this.buffer[i]
        pointer++
      }

      if (this.buffer[i].toString(16) === 'aa') {
        skipNext = true
      }
    }

    debug(
      `Message ${id} with length ${length} arrived (CSUM:${csum === 0}) = `,
      data
    )
    this.push({ data: data.slice(0, length), id, length, crc: csum === 0 })
  }

  _transform(
    chunk: Buffer,
    encoding: string,
    callback: TransformCallback
  ): void {
    debug(
      `Chunk arrived with ${chunk.length} byte, we alreay have ${
        this.byteCount
      }`,
      chunk
    )

    chunk.copy(this.buffer, this.byteCount)
    this.byteCount += chunk.length

    if (
      this.byteCount > 4 &&
      this.buffer.readUInt16BE(2) === this.byteCount - 4
    ) {
      debug('Message complete')
      this.parse()

      this.byteCount = 0
      this.buffer = Buffer.alloc(50)
    }

    callback()
  }
  _flush(callback: TransformCallback): void {
    debug('Flush called')
    callback()
  }
}


src/d302.ts
import * as SerialPort from 'serialport'
import { parsers } from 'serialport'
import * as Debug from 'debug'
import { EventEmitter } from 'events'

import { from, Observable, using, interval, timer } from 'rxjs'
import { mergeMap } from 'rxjs/operators'
import { RFIDReader } from './rfid'

const CMD_INFO = '020100000303'

const CMD_BEEP = '020100000303'

const CMD_READ = '0201a40b0000000000000000000000ac03'

const debug = Debug('ulfalfa:rfid:d302')

export class D302Reader extends RFIDReader {
  constructor(port = '/dev/ttyUSB0') {
    super(port, 4800)
  }

  protected calcCRC(data: Buffer) {
    let csum = 0
    for (let i = 0; i < data.length - 2; i++) {
      // tslint:disable-next-line
      csum = csum ^ data[i]
    }
    return csum
  }

  read(): Promise<string> {
    debug('Atomic Read')
    const timeout = new Promise((resolve, reject) => {
      const id = setTimeout(() => {
        clearTimeout(id)
        this.serialPort.removeAllListeners()
        reject(`Timed out reading code `)
      }, 1000)
    })
    return Promise.race([
      timeout,
      new Promise((resolve, reject) => {
        const cmd = Buffer.from(CMD_READ, 'hex')
        this.serialPort.once('data', (data: Buffer) => {
          debug(`data read`, data)

          if (this.calcCRC(data) !== data[data.length - 2]) {
            throw new Error(`Read error - checksum mismatch`)
          }

          const code = data
            .slice(4, 9)
            .toString('hex')
            .toUpperCase()

          debug('Code read', code)
          resolve(code)
        })
        this.serialPort.write(cmd, (err, result) => {
          if (err) {
            reject(err)
          }
        })
      }),
    ]).then(data => {
      return data as string
    })
  }

  protected atomicWrite(id): Promise<void> {
    debug('Atomic Write', id)
    return new Promise((resolve, reject) => {
      const cmdBuffer = Buffer.from('0201a50b000000000000' + id + 'AA03', 'hex')

      // now calculate checksum and fill cmdBuffer and doubling the 0xAA
      cmdBuffer[cmdBuffer.length - 2] = this.calcCRC(cmdBuffer)

      debug('Command Buffer', cmdBuffer.toString('hex'))
      this.serialPort.once('data', (data: Buffer) => {
        debug('AWrite Returned', data.toString('hex'))
        resolve()
      })

      this.serialPort.write(cmdBuffer, (err, result) => {
        if (err) {
          reject(err)
        }
      })
    })
  }

  getInfo(): Promise<string> {
    const timeout = new Promise((resolve, reject) => {
      const id = setTimeout(() => {
        clearTimeout(id)
        reject(`Timed out reading code `)
      }, 1000)
    })

    return Promise.race([
      timeout,
      new Promise((resolve, reject) => {
        if (!this.serialPort.isOpen) {
          reject('not ready')
          return
        }

        const cmd = Buffer.from(CMD_INFO, 'hex')

        this.serialPort.once('data', data => {
          if (Buffer.compare(data, cmd) === 0) {
            resolve('OK')
          } else {
            reject('This is not a D302 RFID Reader')
          }
        })

        this.serialPort.write(cmd, (err, result) => {
          debug('CMD written', cmd)
          if (err) {
            reject(err)
          }
        })
      }),
    ]).then(data => data as string)
  }
}
